frontend/wasm/src/module/build_ir.rs

```diff
@@ -223,21 +223,18 @@ fn build_data_segments(
     }
 
     // Apply alignment and merging
-    let aligned_segments = align_data_segments(resolved_segments);
-
-    // Now create the data segments in the module builder
-    for segment in aligned_segments {
-        let init = ConstantData::from(segment.data);
+    if let Some(merged_segment) = merge_data_segments(resolved_segments)? {
+        let init = ConstantData::from(merged_segment.data);
         let size = init.len() as u32;
         if let Err(e) = module_builder.define_data_segment(
-            segment.offset,
+            merged_segment.offset,
             init,
-            segment.readonly,
+            merged_segment.readonly,
```

Comment:
We'll need to be really careful with this - we shouldn't merge data segments with mixed `readonly` behavior (i.e. if one segment is `readonly` and the next isn't, we must keep them unmerged).